### ISSUE TITLE
ein:notebook-mode gets disabled inside a code cell when using elpy

### ISSUE THREAD

**Issue by OldIMP** [millejoh/emacs-ipython-notebook#658]
Whenever the cursor is inside a code cell, the _Notebook minor mode_ will disappear and only appear again if the cursor is outside the cell or in a markdown cell.

I know this might be a feature but it makes the notebook unusable since all the keybindings belonging to the Notebook mode are replaced by the elpy ones:
- `C-c C-c` will send the whole buffer to shell interpreter (`elpy-shell-send-region-or-buffer`) instead of eval only the code cell (`ein:worksheet-execute-cell-km`)
- `C-c C-n` will call `elpy-flymake-next-error` instead of going to the next cell (`ein:worksheet-goto-next-input-km`)
- ...

> EIN version is 20200205.1547

> Elpy Configuration
> 
> Emacs.............: 26.3
> Elpy..............: 1.32.0
> Virtualenv........: None
> Interactive Python: jupyter 4.6.1 (c:/Users/shouh/Anaconda3/envs/tf1/Scripts/jupyter.exe)
> RPC virtualenv....: rpc-venv (c:/Users/shouh/AppData/Roaming/.emacs.d/elpy/rpc-venv)
>  Python...........: c:/Users/shouh/Anaconda3/envs/tf1/pythonw.exe 3.7.6 (c:/Users/shouh/Anaconda3/envs/tf1/pythonw.exe)
>  Jedi.............: 0.16.0
>  Rope.............: 0.16.0
>  Autopep8.........: 1.5
>  Yapf.............: 0.29.0
>  Black............: 19.10b0
> Syntax checker....: flake8.exe (c:/Users/shouh/Anaconda3/envs/tf1/Scripts/flake8.exe)

I'm not sure if it's an issue here or in elpy.

Thanks for any fix / workaround!

**Comment by dickmao** [comment]
https://github.com/millejoh/emacs-ipython-notebook/blob/cb2270ae3f445ebe015f6075059312449a1025ff/lisp/poly-ein.el#L16

Could you put a `(message "got here")` on this line?

Working assumption is that it doesn't get there.  In the notebook, `C-h v after-change-major-mode-hook` verify poly-ein--neuter-markdown-mode is there.

**Comment by OldIMP** [comment]
Your assumption is right. The message never got shown.

`poly-ein--neuter-markdown-mode` is in the `after-change-major-mode-hook`:
`(global-magit-file-mode-enable-in-buffers magit-auto-revert-mode-enable-in-buffers git-commit-setup-font-lock-in-buffer polymode-after-change-major-mode-cleanup global-eldoc-mode-enable-in-buffers global-font-lock-mode-enable-in-buffers poly-ein--neuter-markdown-mode)`

**Comment by OldIMP** [comment]
I think `poly-ein--neuter-markdown-mode` was never called, since I added a message at the beginning and it still never got shown

**Comment by OldIMP** [comment]
I've added the message and another one at the beginning of `poly-ein--neuter-markdown-mode` to debug `minor-mode-map-alist`. Following are the messages that I got (some noises were stripped):

```
[yas] Prepared just-in-time loading of snippets successfully.
Turning on magit-auto-revert-mode...done (0.324s, 9 buffers checked)
(caar minor-mode-map-alist): magit-blame-read-only-mode
got here
(caar minor-mode-map-alist): ein:notebook-mode [5 times]
Restoring clock data
(caar minor-mode-map-alist): gnus-dead-summary-mode
got here
Desktop: 1 frame, 5 buffers restored.
(caar minor-mode-map-alist): ein:notebook-mode
(caar minor-mode-map-alist): ein:notebook-mode [2 times]
Formats have changed, recompiling...done
(caar minor-mode-map-alist): ein:notebook-mode
Updating buffer list...done
(caar minor-mode-map-alist): ein:notebook-mode [3 times]
ein: [info] ein:jupyter-server--run: c:/Users/shouh/Anaconda3/envs/tf1/Scripts/jupyter.exe notebook --notebook-dir=c:/dev/handson-ml --no-browser
ein: [info] Login to http://127.0.0.1:8888 complete.
(caar minor-mode-map-alist): ein:notebook-mode
ein: [info] Worksheet exercises_3.ipynb is ready
(caar minor-mode-map-alist): ein:notebook-mode [11 times]
```

**Comment by dickmao** [comment]
That looks right.  To be sure, can you run the following:
```
emacs -Q --batch "Worksheet exercises_3.ipynb" -f package-initialize \
      --eval "(require 'ein)" \
      --eval "(require 'ein-process)" \
      --eval "(setq done-p nil)" \
      --eval "(setq hooked-p nil)" \
      --eval "(cl-loop until done-p \
                  do (ein:process-open-notebook* (expand-file-name buffer-file-name) \
                        (lambda (nb &rest _args) \
                           (with-current-buffer (ein:notebook-buffer nb) \
                               (goto-char (point-min)) \
                               (cl-assert (string= major-mode \"fundamental-mode\")) \
                               (princ (format \"%s %s \n\" major-mode (buffer-name))) \
                               (cl-assert (memq (function poly-ein--neuter-markdown-mode) after-change-major-mode-hook)) \
                               (add-hook (quote after-change-major-mode-hook) (lambda () (setf hooked-p t)) t) \
                               (call-interactively (function ein:worksheet-goto-next-input)) \
                               (pm-switch-to-buffer) \
                               (princ (format \"%s %s\n\" major-mode (buffer-name))) \
                               ) \
                           (setq done-p t))) \
                  do (sleep-for 0 3000) \
                  finally (princ (format \"all %s\n\" (if hooked-p \"good\" \"bad\"))))"
```

When I run it, I get something like:
```
ein: [info] Login to http://127.0.0.1:8888 complete.
ein: [info] Worksheet Untitled.ipynb is ready
Can’t guess python-indent-offset, using defaults: 4
fundamental-mode *ein: http://127.0.0.1:8888/Untitled.ipynb* 
ein:markdown-mode  *ein: http://127.0.0.1:8888/Untitled.ipynb*[ein:markdown]

ein: [info] Notebook Untitled.ipynb is ready
all good
```

**Comment by OldIMP** [comment]
```
ein: [info] Login to http://127.0.0.1:8888 complete.
(caar minor-mode-map-alist): magit-popup-help-mode
got here
ein: [info] Worksheet exercises_3.ipynb is ready
(caar minor-mode-map-alist): ein:notebook-mode
(caar minor-mode-map-alist): outline-minor-mode
got here
ein: [info] Notebook exercises_3.ipynb is already open
fundamental-mode *ein: http://127.0.0.1:8888/exercises_3.ipynb*
python-mode  *ein: http://127.0.0.1:8888/exercises_3.ipynb*[python]
fundamental-mode *ein: http://127.0.0.1:8888/exercises_3.ipynb*
python-mode  *ein: http://127.0.0.1:8888/exercises_3.ipynb*[python]

ein: [info] Notebook exercises_3.ipynb is ready
(caar minor-mode-map-alist): ein:notebook-mode
(caar minor-mode-map-alist): ein:notebook-mode
all good

ein: [info] Signaled ein server with pid 17972
```

**Comment by dickmao** [comment]
Try this then:
```
emacs "exercises_3.ipynb" \
--eval "(defun oldimp-print () \
            (princ (format \"%s %s\n\" major-mode (buffer-name)) (function external-debugging-output)) \
            (mapc (lambda (x) (when x (princ (format \"%s: %s\n\" (symbol-name x) (symbol-value x)) (function external-debugging-output)))) \
              (mapcar (function intern-soft) (split-string \"overriding-terminal-local-map overriding-local-map emulation-mode-map-alists minor-mode-overriding-map-alist\"))) \
            (princ (format \"char-prop: %s\" (get-char-property (point) (quote keymap))) (function external-debugging-output)) \
            (princ (format \"local-map: %s\" (get-text-property (point) (quote local-map))) (function external-debugging-output)) \
            (princ (format \"current-local-map: %s\" (current-local-map)) (function external-debugging-output)) \
            (princ (format \"current-global-map: %s\" (current-global-map)) (function external-debugging-output)))" > oldimp-print 2>&1
```

Then `C-c C-o` exercises_3.ipynb.  Move the cursor to the python cell.  Invoke `M-: (oldimp-print)`.  Exit emacs.  Look for "elpy" in the file `oldimp-print`.

**Comment by OldIMP** [comment]
Sry but "elpy" is not in it:
[oldimp-print.txt](https://github.com/millejoh/emacs-ipython-notebook/files/4177686/oldimp-print.txt)

**Comment by dickmao** [comment]
I'm afraid then i cannot help you.  If you wish to help me, you can try `emacs -Q -f package-initialize`, which bypasses any of your personal configuration, and report here that you can reproduce the error.  Then call M-x ein:dev-bug-report-template so I can gather your OS details.  Thanks.

**Comment by OldIMP** [comment]
## Problem description

_ein:notebook-mode_ gets disabled inside a code cell when using _elpy_

## Steps to reproduce the problem

0. `emacs -Q -f package-initialize`
1. install _elpy_ & `(elpy-enable)`
2. open a Jupyter notebook and go to a code cell
3. Notebook minor mode disappears



## System info:

```cl
("EIN system info"
 :emacs-version "GNU Emacs 26.3 (build 1, x86_64-w64-mingw32)
 of 2019-08-29"
 :emacs-bzr-version "96dd0196c28bc36779584e47fffcca433c9309cd"
 :window-system w32
 :emacs-variant nil
 :os
 (:uname nil
	 :lsb-release nil)
 :notebook
 "Name: notebook
Version: 6.0.3
Summary: A web-based notebook environment for interactive computing
Home-page: http://jupyter.org
Author: Jupyter Development Team
Author-email: jupyter@googlegroups.com
License: BSD
Location: c:\\users\\shouh\\anaconda3\\envs\\tf1\\lib\\site-packages
Requires: ipython-genutils, pyzmq, ipykernel, tornado, nbformat, traitlets, Send2Trash, terminado, prometheus-client, jupyter-core, jupyter-client, jinja2, nbconvert
Required-by: widgetsnbextension, nbdime, jupyterlab, jupyterlab-server
"
 :ipython "7.10.1
"
 :image-types
 (svg png gif tiff jpeg xpm xbm pbm)
 :image-types-available
 (svg png gif tiff jpeg xpm xbm pbm)
 :request
 (:backend curl)
 :ein
 (:version "20200209.35"
	   :source-dir "c:/Users/shouh/AppData/Roaming/.emacs.d/elpa/ein-20200209.35/")
 :lib
 ((:name "websocket"
	 :path "~/.emacs.d/elpa/websocket-20200102.637/websocket.elc"
	 :featurep t
	 :version-var websocket-version
	 :version "1.12")
  (:name "anaphora"
	 :path "~/.emacs.d/elpa/anaphora-20180618.2200/anaphora.elc"
	 :featurep t
	 :version-var nil
	 :version nil)
  (:name "request"
	 :path "~/.emacs.d/elpa/request-20200127.21/request.elc"
	 :featurep t
	 :version-var request-version
	 :version "0.3.0")
  (:name "deferred"
	 :path "~/.emacs.d/elpa/deferred-20170901.1330/deferred.elc"
	 :featurep t
	 :version-var deferred:version
	 :version "0.5.0")
  (:name "polymode"
	 :path "~/.emacs.d/elpa/polymode-20191208.1239/polymode.elc"
	 :featurep t
	 :version-var nil
	 :version nil)
  (:name "kubernetes"
	 :path "~/.emacs.d/elpa/kubernetes-20200114.436/kubernetes.elc"
	 :featurep t
	 :version-var nil
	 :version nil)
  (:name "dash"
	 :path "~/.emacs.d/elpa/dash-20200119.2310/dash.elc"
	 :featurep t
	 :version-var nil
	 :version nil)
  (:name "s"
	 :path "~/.emacs.d/elpa/s-20180406.808/s.elc"
	 :featurep t
	 :version-var nil
	 :version nil)))
```

**Comment by dickmao** [comment]
Thank you, and wait.... I was able to reproduce the problem.  Can you install the use-package package and initiate elpy as follows?
```
(use-package elpy
  ;; emacs.stackexchange.com/questions/10065
  :commands elpy-enable
  :init (with-eval-after-load 'python (elpy-enable))
  :config
  (setq elpy-rpc-backend "jedi"
        elpy-rpc-project-specific 't)
  (delete 'elpy-module-highlight-indentation elpy-modules)
  (delete 'elpy-module-flymake elpy-modules))
```

**Comment by dickmao** [comment]
Thank you for reporting this bug and staying with it.

I was able to finally reproduce the error with `emacs -Q -f package-initialize --eval "(setq debug-on-error t)" --eval "(elpy-enable)"`.  Sorry I should have tried this much earlier -- I had assumed it was a keymap issue, and not a general bug.

I hope the above PR fixes.  Thanks again.

**Comment by OldIMP** [comment]
Verified that bug was fixed in 20200210.452.

Thx for the professional work! This fixed lots of other problems w/ elpy, too, e.g. now the notebook is opened in the front ground.
